### ISSUE TITLE
Basic asan support

### ISSUE THREAD

**Issue by luisggpina** [daniel-grumberg/varan-div#3]
I was adding basic asan support to Varan and I'd like to write the following rules:

```
rt_sigaction(sig,_,_) { $(sig) == SIGFPE || $(sig) == SIGBUS || $(sig) == SIGSEGV }  => Nothing

Nothing => sched_setaffinity(_,_,_) 
```

The first one skips all `rt_sigaction` leader syscalls for some signals (`SIGFPE`, `SIGBUS`, and `SIGSEGV`).

The second skips all `sched_setaffinity` that the follower issues.

Does the current language supports these?  Am I specifying them wrong?

**Comment by daniel-grumberg** [comment]
No, the DSL does not support this at the moment. I will think about some appropriate syntax. But, I already have a few ideas ...

**Comment by daniel-grumberg** [comment]
Just give until next Monday, I had a technical interview on Wednesday, I also need to complete a coding assignment due on Friday for another application and I am attending a careers fair here in Luxembourg on Friday as well. I will get to it by Monday though. So far for clarity's sake, I think skipping syscalls on the follower no matter what deserves special syntax.

**Comment by luisggpina** [comment]
Ok, that sounds good.

I might open another issue for features I'd like to have to support Valgrind divergences, given that I'm now working on that.  The timeline you suggested is fine, please don't interpret a new issue as increased urgency on this project.

**Comment by daniel-grumberg** [comment]
Ok, are we skipping system call on the follower whenever they appear or are there any times we do it only if a set of conditions are met ?

**Comment by luisggpina** [comment]
Hmm.  For now I only need to skip one syscall without any conditions.  But I wonder if adding a guard, like on the line above, would make it much more difficult to implement.

**Comment by daniel-grumberg** [comment]
what I was thinking was specialized syntax for the skip no matter what type of cases, something along the lines of:
```
skip = [sched_setaffinity]
```
and for guard type cases to include them in a normal rewrite sequence but to add some syntax to say it should be skipped:
```
lhs(_,_,_) => sched_setaffinity(a,b,c) skip,
                    open(_,_,_)
```

**Comment by daniel-grumberg** [comment]
I will get started with the first one tonight as I have been really busy with my placement applications

**Comment by luisggpina** [comment]
I like your idea, both approaches work for this example and look like they will work for more complex examples.  I'll try to take a quick look at Valgrind in the meantime and report back if I find something that requires more support.

**Comment by luisggpina** [comment]
I think the case `Nothing => syscall` does need a guard.  For instance, when a program under Valgrind mmap's a file, Valgrind performs extra syscalls.  The leader never registers any `mmap` in the ringbuffer.

So, this is what a rule could look like:

```
Nothing => mmap(_,_,_,flags,_) { ((flags & MAP_ANONYMOUS) == 0) },
           fstat(),
           readlink(),
           stat()

Nothing => mmap(_,_,_,_,_)
```

The first line on the first rule acts as a guard.  If the first rule doesn't trigger, the second rule does.

**Comment by daniel-grumberg** [comment]
I see, but would you not expect a `mmap` call on the leader as well then ? If that is the case that works fine in the current language:
```
mmap(_,_,_,flags,_) { return ((flags & MAP_ANONYMOUS) == 0); }
   => mmap(_,_,_,flags,_) { return ((flags & MAP_ANONYMOUS) == 0);},
         fstat(),
         readlink(),
         stat()

mmap(_,_,_,_,_) => self
```

The first rule says that if we have that particular mmap on the leader we should expect it on the follower along with the other 3 system calls. The second rule is the catchall rule that says mmap is mapped to itself, which I believe is not even necessary as the default case should handle it.

**Comment by luisggpina** [comment]
The leader never registers any `mmap` calls on the ringbuffer.

**Comment by luisggpina** [comment]
This is now working and I'm marking this as fixed by PRs #5 #6 #8
